docs/src/adr/004adr-annotations.md

```diff
@@ -0,0 +1,136 @@
+ 1. Recursive operators, that are declared either at the top-level or by using
+    LET-IN.
+
+For an example, see Section 3.
+
+
+## 2. Annotations syntax
+
+An annotation is a string that follows the grammar (question mark denotes
+optional rules):
+
+```
+Annotation  ::= '@' javaIdentifier ( '(' ArgList? ')' )?
+ArgList     ::= (Arg) ( ',' Arg )*
+Arg         ::= (string | integer | boolean)
+string      ::= '"' <char sequence> '"'
+integer     ::= '-'? [0-9]+
+boolean     ::= ('false' | 'true')
+```
+
+Java Language Specification defines how a [Java identifier] looks like.
+The sequence `<char sequence>` is a sequence admitted by [JavaTokenParsers]:
+  
+  - Any character except double quotes, control characters or backslash `\`
+  - A backslash followed by another backslash, a single or double quote,
+    or one of the letters `b`, `f`, `n`, `r` or `t`
+  - `\` followed by u followed by four hexadecimal digits
+  
+**Examples.** The following strings are examples of syntactically correct
+annotations:
+
+ - `@tailrec`
+ - `@type("(Int, Int) => Int")`
+ - `@random(true)`
+ - `@deprecated("Use operator Foo instead")`
+ - `@range(0, 100)`
+
+The above examples are just syntactically correct. They meaning, if there is
+any, is defined by the tool that is reading these annotations.
+
+## 3. An annotated specification
+
+The following specification shows how to write annotations, so they can be
+correctly parsed by the SANY parser and Apalache. Note the location of comments
+in front of: local operators, LET-definitions, and recursive operators.
+Although these locations may seem to be suboptimal, this is how the SANY
+parser locates comments that precede declarations.
+
+```tla
+-------- MODULE Annotations ----------
+EXTENDS Integers
+
+CONSTANT
+  \* @type(" Int ")
+  N
+
+VARIABLE
+  \* @type(" Set(Int) ")
+  set
+
+\* @pure
+\* @type(" Int => Int ")
+Inc(n) == n + 1
+
+\* @type(" Int => Int ")
+LOCAL LocalInc(x) == x + 1
+
+A(n) ==
+  LET \* @pure
+      \* @type(" Int => Int ")
+      Dec(x) == x + 1
+  IN
+  Dec(n)
+
+RECURSIVE Fact(_)
+\* @tailrec
+\* @type(" Int => Int ")
+Fact(n) ==
+  IF n <= 1 THEN 1 ELSE n * Fact(n - 1)
+
+\* @tailrec
+\* @type(" Int -> Int ")
+FactFun[n \in Int] ==
+  IF n <= 1 THEN 1 ELSE n * FactFun[n - 1]
+
+======================================
+```
+
+## 4. Implementation
+
+The implementation of the annotation parser can be found in the class
+`at.forsyte.apalache.io.annotations.AnnotationParser` of the module
+`tla-import`. (TODO: place a link when the PR is merged).
+
+## 5. Discussion
+
+Most likely, this topic does not deserve much discussion, as we are using
+the pretty standard syntax of Java annotations. So we are following the
+principle of the least surprise.
+
+A more concise syntax for type annotations could be as follows:
+
+```tla
+\* @type: Int => Int
+Inc(n) == n + 1
+```
+
+However, this concise syntax is technically harder to parse (SANY is prunning
```

Comment:
How much harder to parse?

Reply:
See the answer above. I was about to do it and then realized that it is nearly impossible with the current AST that we get from SANY